survey/forms.py

```diff
@@ -52,20 +56,59 @@ def __init__(self, *args, **kwargs):
             else:
                 self.add_question(question, data)
 
+        self._get_preexisting_response()
+
+        if not self.survey.editable_answers and self.response is not None:
+            for name in self.fields.keys():
+                self.fields[name].widget.attrs["disabled"] = True
+
     def _get_preexisting_response(self):
         """ Recover a pre-existing response in database.
 
-        The user must be logged.
+        The user must be logged. Will store the response retrieved in an attribute
+        to avoid multiple db calls.
+
         :rtype: Response or None"""
+        if self.response:
+            return self.response
+
         if not self.user.is_authenticated:
-            return None
+            self.response = None
+        else:
+            try:
+                self.response = Response.objects.prefetch_related("user", "survey").get(
+                    user=self.user, survey=self.survey
+                )
+            except Response.DoesNotExist:
+                LOGGER.debug(
+                    "No saved response for '%s' for user %s", self.survey, self.user
+                )
+                self.response = None
+        return self.response
+
+    def _get_preexisting_answers(self):
+        """ Recover pre-existing answers in database.
+
+        The user must be logged. A Response containing the Answer must exists.
+        Will create an attribute containing the answers retrieved to avoid multiple
+        db calls.
+
+        :rtype: dict of Answer or None"""
+        if self.answers:
+            return self.answers
+
+        response = self._get_preexisting_response()
+        if response is None:
+            self.answers = None
         try:
-            return Response.objects.get(user=self.user, survey=self.survey)
-        except Response.DoesNotExist:
-            LOGGER.debug(
```

Comment:
I think having debug logging can be useful, we should probably keep those.

Reply:
Indeed but the try/except is still done in the `_get_preexisting_response` method, just above. :)

Reply:
Ok then :)